File: Solutions/380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.cs
using System;
using System.Collections.Generic;

public class RandomizedSet
{
    // Dictionary to store value → index in the list
    private Dictionary<int, int> dict;
    
    // List to store the values for O(1) random access
    private List<int> list;
    
    // Random number generator for getRandom
    private Random rand;

    // Constructor: initialize data structures
    public RandomizedSet()
    {
        dict = new Dictionary<int, int>(); // value → index mapping
        list = new List<int>();            // list of values
        rand = new Random();               // random number generator
    }

    // Inserts a value to the set. Returns true if the set did not already contain the specified element
    public bool Insert(int val)
    {
        // If val already exists in dictionary, return false
        if (dict.ContainsKey(val))
            return false;

        // Add val to the end of the list
        list.Add(val);

        // Store the index of val in the dictionary
        dict[val] = list.Count - 1;

        // Successfully inserted
        return true;
    }

    // Removes a value from the set. Returns true if the set contained the specified element
    public bool Remove(int val)
    {
        // If val does not exist in dictionary, return false
        if (!dict.ContainsKey(val))
            return false;

        // Get index of the element to remove
        int idxToRemove = dict[val];

        // Get the last element in the list
        int lastElement = list[list.Count - 1];

        // Move the last element to the position of the element to remove
        list[idxToRemove] = lastElement;

        // Update the dictionary for the moved element
        dict[lastElement] = idxToRemove;

        // Remove the last element from the list
        list.RemoveAt(list.Count - 1);

        // Remove the value from the dictionary
        dict.Remove(val);

        // Successfully removed
        return true;
    }

    // Get a random element from the set
    public int GetRandom()
    {
        // Generate a random index from 0 to list.Count - 1
        int idx = rand.Next(list.Count);

        // Return the value at that random index
        return list[idx];
    }
}

/**
 * Example usage:
 * RandomizedSet obj = new RandomizedSet();
 * Console.WriteLine(obj.Insert(1));   // true
 * Console.WriteLine(obj.Remove(2));   // false
 * Console.WriteLine(obj.Insert(2));   // true
 * Console.WriteLine(obj.GetRandom()); // 1 or 2
 * Console.WriteLine(obj.Remove(1));   // true
 * Console.WriteLine(obj.Insert(2));   // false
 * Console.WriteLine(obj.GetRandom()); // 2
 */

File: Solutions/380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.swift
class RandomizedSet {

    // Array to store elements for O(1) random access
    private var arr: [Int]
    
    // Dictionary to store value -> index mapping
    private var map: [Int: Int]
    
    /** Initialize your data structure here. */
    init() {
        arr = []
        map = [:]
    }
    
    /**
     * Inserts a value to the set.
     * Returns true if the set did not already contain the specified element.
     */
    func insert(_ val: Int) -> Bool {
        
        // If value already exists, do not insert
        if map[val] != nil {
            return false
        }
        
        // Append value to array
        arr.append(val)
        
        // Store its index in the map
        map[val] = arr.count - 1
        
        return true
    }
    
    /**
     * Removes a value from the set.
     * Returns true if the set contained the specified element.
     */
    func remove(_ val: Int) -> Bool {
        
        // If value does not exist, return false
        guard let index = map[val] else {
            return false
        }
        
        // Get the last element in the array
        let lastElement = arr[arr.count - 1]
        
        // Swap the element to remove with the last element
        arr[index] = lastElement
        
        // Update the index of the last element in the map
        map[lastElement] = index
        
        // Remove the last element from the array
        arr.removeLast()
        
        // Remove the value from the map
        map.removeValue(forKey: val)
        
        return true
    }
    
    /**
     * Get a random element from the set.
     * Each element must have the same probability of being returned.
     */
    func getRandom() -> Int {
        
        // Generate a random index in the array
        let randomIndex = Int.random(in: 0..<arr.count)
        
        // Return the element at that index
        return arr[randomIndex]
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * let obj = RandomizedSet()
 * let ret_1: Bool = obj.insert(val)
 * let ret_2: Bool = obj.remove(val)
 * let ret_3: Int = obj.getRandom()
 */